File: 백준/Silver/1149. RGB거리/RGB거리.py
import sys

input = sys.stdin.readline


n = int(input())
data = []
for _ in range(n):
    r,g,b = map(int, input().split())
    data.append((r,g,b))

#모든 집을 칠하는 비용의 최솟값
#dp 테이블에는 각 색깔을 칠했을 떄의 최솟값을 저장하자.
dp = [[0] * 3 for _ in range(n)] 
dp[0][0] = data[0][0]
dp[0][1] = data[0][1]
dp[0][2] = data[0][2] #자명한건 미리 초기화

for i in range(1,n): #각 색깔별로 구해야.
    #현재 j번째 색을 칠할 때, 최솟값을 저장해야 하니까 아래와 같이 작성해야함.
    dp[i][0] = min(dp[i-1][1] + data[i][0], dp[i-1][2] + data[i][0]) 
    dp[i][1] = min(dp[i-1][0] + data[i][1], dp[i-1][2] + data[i][1])
    dp[i][2] = min(dp[i-1][0] + data[i][2], dp[i-1][1] + data[i][2])
    
res = min(dp[n-1][0], dp[n-1][1], dp[n-1][2])

print(res)

File: 백준/Silver/1149. RGB거리/RGB거리.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int n;
    static int red = 0, green = 1, blue = 2;
    static int[][] dp, data;  // ! 2차원인 이유 : 각 집마다 3개의 값 가질 수 있음 : R,G,B
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        dp = new int[n+1][3]; // ! 1번 인덱스부터 사용
        data = new int[n+1][3];
        for (int i = 1; i <= n; i++) {
            st = new StringTokenizer(br.readLine());
            int r = Integer.parseInt(st.nextToken());
            int g = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            data[i][red] = r;
            data[i][green] = g;
            data[i][blue] = b;
        }

        dp[1][red] = data[1][red];
        dp[1][green] = data[1][green];
        dp[1][blue] = data[1][blue];
        // ! 자명한 값 초기화

        // ! dp[i][j] : i번째 집의 j번째 색(r,g,b)을 칠할 때의 최소값
        for (int i = 2; i <= n; i++) {
            dp[i][red] = Math.min(dp[i - 1][green], dp[i - 1][blue]) + data[i][red];
            dp[i][green] = Math.min(dp[i - 1][red], dp[i - 1][blue]) + data[i][green];
            dp[i][blue] = Math.min(dp[i - 1][red], dp[i - 1][green]) + data[i][blue];
        }

        int min = Integer.MAX_VALUE;
        for (int i = 0; i < 3; i++) {
            min = Math.min(dp[n][i], min);
        }
        System.out.println(min);
    }
}